src/StableBuffer.mo

```diff
@@ -86,88 +89,95 @@ module {
   };
 
   /// Returns the count of elements in the buffer
-  public func size<X>(buffer: StableBuffer<X>) : Nat { buffer.count };
+  public func size<X>(buffer : StableBuffer<X>) : Nat { buffer.count };
 
   /// Resets the buffer.
-  public func clear<X>(buffer: StableBuffer<X>): () {
+  public func clear<X>(buffer : StableBuffer<X>) : () {
     buffer.count := 0;
   };
 
   /// Returns a copy of this buffer.
-  public func clone<X>(buffer: StableBuffer<X>) : StableBuffer<X> {
+  public func clone<X>(buffer : StableBuffer<X>) : StableBuffer<X> {
     let c = initPresized<X>(buffer.elems.size());
-    var i = 0;
+    var i = vals(buffer);
+
     label l loop {
-      if (i >= buffer.count) break l;
-      add(c, buffer.elems[i]);
-      i += 1;
+      switch (i.next()) {
+        case null break l;
+        case (?x) { add(c, x) };
+      };
     };
-    c
+
+    c;
   };
 
   /// Returns an `Iter` over the elements of this buffer.
-  public func vals<X>(buffer: StableBuffer<X>) : { next : () -> ?X } = object {
+  public func vals<X>(buffer : StableBuffer<X>) : { next : () -> ?X } = object {
     var pos = 0;
     public func next() : ?X {
       if (pos == buffer.count) { null } else {
-        let elem = ?buffer.elems[pos];
+        let elem = buffer.elems[pos];
         pos += 1;
-        elem
-      }
-    }
+        elem;
+      };
+    };
   };
 
   /// Creates a Buffer from an Array
-  public func fromArray<X>(xs: [X]): StableBuffer<X> {
-    let ys: StableBuffer<X> = initPresized(xs.size());
+  public func fromArray<X>(xs : [X]) : StableBuffer<X> {
+    let ys : StableBuffer<X> = initPresized(xs.size());
     for (x in xs.vals()) {
       add(ys, x);
     };
 
-    ys
+    ys;
   };
 
   /// Creates a new array containing this buffer's elements.
-  public func toArray<X>(buffer: StableBuffer<X>) : [X] =
-    // immutable clone of array
-    Prim.Array_tabulate<X>(
-      buffer.count,
-      func(x : Nat) : X { buffer.elems[x] }
-    );
+  public func toArray<X>(buffer : StableBuffer<X>) : [X] =
```

Comment:
Here's another instance where if the developer sets one of the elements with an `index < count` to null, this will trap when copying it over to an array. This trap will happen at runtime instead of at compile time, and feels a bit unsafe-ish/buggy to me.

If we make these changes, then I definitely need to start including usage warnings all over the code 😅

Reply:
I tried this example in the [motoko playground](https://m7sm4-2iaaa-aaaab-qabra-cai.raw.ic0.app/?tag=2912498390). If you notice the `say()` fn doesn't trap even though you enter a `null` value. This means that Motoko supports optional null values `?null` 💪 .

Reply:
> "This means that Motoko supports optional null values"

🤯